sdk/python/tests/test_RuleBasedTransactionFactory.py

```diff
@@ -0,0 +1,488 @@
+from symbolchain.CryptoTypes import Hash256, PublicKey
+from symbolchain.RuleBasedTransactionFactory import RuleBasedTransactionFactory
+
+# region Module
+
+
+class Module:
+	EnumAlias = Enum
+	FlagAlias = Flag
+	BaseValueAlias = BaseValue
+
+	SigningPublicKey = PublicKey
+
+	class Hash256(ByteArray):
+		def __init__(self, hash256):
+			super().__init__(32, hash256, Module.Hash256)
+
+	class MosaicFlags(Flag):
+		NONE = 0
+		SUPPLY_MUTABLE = 1
+		TRANSFERABLE = 2
+		RESTRICTABLE = 4
+		REVOKABLE = 8
+
+	class NetworkType(Enum):
+		MAINNET = 104
+		TESTNET = 152
+
+	class UnresolvedMosaicId(BaseValue):
+		def __init__(self, unresolved_mosaic_id=0):
+			super().__init__(8, unresolved_mosaic_id, Module.UnresolvedMosaicId)
+
+	class Amount(BaseValue):
+		def __init__(self, amount=0):
+			super().__init__(8, amount, Module.Amount)
+
+	class UnresolvedMosaic:
+		TYPE_HINTS = {
+			'mosaic_id': 'pod:UnresolvedMosaicId',
+			'amount': 'pod:Amount'
+		}
+
+		def __init__(self):
+			self.mosaic_id = Module.UnresolvedMosaicId()
+			self.amount = Module.Amount()
+
+	class StructPlain:
+		TYPE_HINTS = {}
+
+		def __init__(self):
+			self.mosaic_id = 0
+			self.amount = 0
+
+	class StructArrayMember:
+		TYPE_HINTS = {
+			'mosaic_ids': 'array[UnresolvedMosaicId]'
+		}
+
+		def __init__(self):
+			self.mosaic_ids = []
+
+	class StructEnumMember:
+		TYPE_HINTS = {
+			'network': 'enum:NetworkType'
+		}
+
+		def __init__(self):
+			self.network = None
+
+	class StructStructMember:
+		TYPE_HINTS = {
+			'mosaic': 'struct:UnresolvedMosaic'
+		}
+
+		def __init__(self):
+			self.mosaic = None
+
+	class StructHashMember:
+		TYPE_HINTS = {
+			'hash': 'Hash256'
+		}
+
+		def __init__(self):
+			self.hash = None
+
+
+# endregion
+
+class RuleBasedTransactionFactoryTest(unittest.TestCase):
+	# pylint: disable=too-many-public-methods
+
+	# region pod parser
+
+	def _assert_pod_parser(self, input_value, expected_value, typing_rules=None):
+		# Arrange:
+		factory = RuleBasedTransactionFactory(Module, None, typing_rules or {})
+		factory.add_pod_parser('SigningPublicKey', PublicKey)
+		rule = factory.rules['SigningPublicKey']
+
+		# Act:
+		parsed = rule(input_value)
+
+		# Assert:
+		self.assertEqual(expected_value, parsed)
+
+	def test_pod_parser_can_handle_raw_value(self):
+		self._assert_pod_parser(
+			'364F3694A022DB4DC59558944707C6679F6FD7E1A7B99CDE8F7D16D3FF515D28',
+			PublicKey('364F3694A022DB4DC59558944707C6679F6FD7E1A7B99CDE8F7D16D3FF515D28'))
+
+	def test_pod_parser_can_handle_typed_value(self):
+		public_key = PublicKey('364F3694A022DB4DC59558944707C6679F6FD7E1A7B99CDE8F7D16D3FF515D28')
+		self._assert_pod_parser(public_key, public_key)
+
+	def test_pod_parser_uses_type_rule_override_when_available(self):
+		self._assert_pod_parser(
+			PublicKey('364F3694A022DB4DC59558944707C6679F6FD7E1A7B99CDE8F7D16D3FF515D28'),
+			'pubkey 364F3694A022DB4DC59558944707C6679F6FD7E1A7B99CDE8F7D16D3FF515D28',
+			{PublicKey: lambda value: f'pubkey {value}'})
```

Comment:
one more test for string-based overrides?
(i.e. via 'Amount' and `class UnresolvedMosaic`)

Reply:
now it is only a type lookup, not a type name lookup too